File: KhmerDictionary/Sources/Updates/ManifestSecurity.swift
import CryptoKit
import Foundation

enum ManifestSecurity {
    static func buildSigningPayload(_ manifest: DatabaseUpdateManifest) -> String {
        [
            String(manifest.schemaVersion),
            String(manifest.release.versionCode),
            manifest.release.versionName,
            manifest.release.publishedAt,
            manifest.database.url,
            manifest.database.sha256.lowercased(),
            String(manifest.database.sizeBytes)
        ].joined(separator: "\n")
    }

    static func verifySignature(manifest: DatabaseUpdateManifest, publicKeyHex: String) -> Bool {
        do {
            let signature = try Data(hex: manifest.database.signatureHex)
            let publicKeyData = try Data(hex: publicKeyHex)
            guard signature.count == 64, publicKeyData.count == 32 else {
                return false
            }

            let publicKey = try Curve25519.Signing.PublicKey(rawRepresentation: publicKeyData)
            let signedPayload = Data(buildSigningPayload(manifest).utf8)
            return publicKey.isValidSignature(signature, for: signedPayload)
        } catch {
            return false
        }
    }

    static func isNewerRelease(manifest: DatabaseUpdateManifest, currentVersionCode: Int?) -> Bool {
        guard let currentVersionCode else {
            return true
        }

        return manifest.release.versionCode > currentVersionCode
    }

    static func sha256Hex(for data: Data) -> String {
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

private enum HexError: Error {
    case invalidLength
    case invalidCharacters
}

private extension Data {
    init(hex: String) throws {
        guard hex.count % 2 == 0 else {
            throw HexError.invalidLength
        }

        let normalized = hex.lowercased()
        guard normalized.range(of: "^[0-9a-f]+$", options: .regularExpression) != nil else {
            throw HexError.invalidCharacters
        }

        var data = Data(capacity: normalized.count / 2)
        var index = normalized.startIndex
        while index < normalized.endIndex {
            let nextIndex = normalized.index(index, offsetBy: 2)
            let byteString = normalized[index..<nextIndex]
            guard let byte = UInt8(byteString, radix: 16) else {
                throw HexError.invalidCharacters
            }

            data.append(byte)
            index = nextIndex
        }

        self = data
    }
}

File: src/features/updates/manifest-security.ts
import nacl from 'tweetnacl';

export type DatabaseUpdateManifest = {
  schemaVersion: 1;
  release: {
    versionCode: number;
    versionName: string;
    publishedAt: string;
  };
  database: {
    url: string;
    sha256: string;
    sizeBytes: number;
    signatureHex: string;
  };
};

function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new Error('hex string must have an even length');
  }
  const normalized = hex.toLowerCase();
  if (!/^[0-9a-f]+$/.test(normalized)) {
    throw new Error('invalid hex string');
  }

  const bytes = new Uint8Array(normalized.length / 2);
  for (let i = 0; i < normalized.length; i += 2) {
    bytes[i / 2] = Number.parseInt(normalized.slice(i, i + 2), 16);
  }

  return bytes;
}

export function buildManifestSigningPayload(manifest: DatabaseUpdateManifest): string {
  return [
    String(manifest.schemaVersion),
    String(manifest.release.versionCode),
    manifest.release.versionName,
    manifest.release.publishedAt,
    manifest.database.url,
    manifest.database.sha256.toLowerCase(),
    String(manifest.database.sizeBytes),
  ].join('\n');
}

export async function verifyManifestSignature(
  manifest: DatabaseUpdateManifest,
  publicKeyHex: string
): Promise<boolean> {
  try {
    const payload = buildManifestSigningPayload(manifest);
    const signatureBytes = hexToBytes(manifest.database.signatureHex);
    const publicKeyBytes = hexToBytes(publicKeyHex);
    if (signatureBytes.length !== 64 || publicKeyBytes.length !== 32) {
      return false;
    }

    return nacl.sign.detached.verify(
      new TextEncoder().encode(payload),
      signatureBytes,
      publicKeyBytes
    );
  } catch {
    return false;
  }
}

export function isNewerRelease(
  manifest: Pick<DatabaseUpdateManifest, 'release'>,
  currentVersionCode: number | null
): boolean {
  if (currentVersionCode == null) {
    return true;
  }

  return manifest.release.versionCode > currentVersionCode;
}
